tests/Utility/BigIntegerTest.php

```diff
@@ -0,0 +1,272 @@
+<?php
+/*
+ * This file is part of the PHPASN1 library.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FG\Test\Utility;
+
+use FG\Utility\BigInteger;
+use PHPUnit\Framework\TestCase;
+
+abstract class BigIntegerTest extends TestCase
+{
+    /**
+     * Whether the current testing instance supports this implementation.
+     * @return bool
+     */
+    abstract protected function _isSupported();
+
+    /**
+     * Return mode expected by BigInteger::setPrefer.
+     * @return string
+     */
+    abstract protected function _getMode();
+
+    protected function setUp()
+    {
+        if (!$this->_isSupported()) {
+            $this->markTestSkipped(sprintf('Mode %s is not supported.', $this->_getMode()));
+            return;
+        }
+
+        BigInteger::setPrefer($this->_getMode());
+    }
+
+
+
+    public function testCreateFromString()
+    {
+        $a = BigInteger::create('8888');
+        $this->assertSame('8888', (string)$a);
+
+        $a = BigInteger::create('0');
+        $this->assertSame('0', (string)$a);
+
+        $a = BigInteger::create('-8888');
+        $this->assertSame('-8888', (string)$a);
+
+        $a = BigInteger::create('18446744073709551616');
+        $this->assertSame('18446744073709551616', (string)$a);
+    }
+
+    public function testCreateFromInteger()
+    {
+        $a = BigInteger::create(8888);
+        $this->assertSame('8888', (string)$a);
+
+        $a = BigInteger::create(0);
+        $this->assertSame('0', (string)$a);
+
+        $a = BigInteger::create(-8888);
+        $this->assertSame('-8888', (string)$a);
+
+        $a = BigInteger::create(PHP_INT_MAX);
+        $this->assertSame((string)PHP_INT_MAX, (string)$a);
+    }
+
+    public function testToInteger()
+    {
+        $a = BigInteger::create(8888);
+        $this->assertSame(8888, $a->toInteger());
+
+        $a = BigInteger::create(0);
+        $this->assertSame(0, $a->toInteger());
+
+        $a = BigInteger::create(-8888);
+        $this->assertSame(-8888, $a->toInteger());
+
+        $a = BigInteger::create(PHP_INT_MAX);
```

Comment:
What happens if you do this?

$a = BigInteger::create('123897123987123971293871298371293918237812739127398');

It's just as likely a codepath could use toInteger on such a number - maybe implementations should refuse if the value is above PHP_INT_MAX (an exception)?

Reply:
I think the `gmp_intval` truncates to `PHP_INT_MAX`. I am of two minds about it.

Originally, I figured that it is up to the caller to only as for an integer when it can be represented. I added some checks and will now throw an exception, but such checks may be overkill and introduce a performance hit.